In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import warnings
import copy
# warnings.filterwarnings('ignore')
plt.rcParams['figure.dpi'] = 80

### Эксперимент

#### Фиксированные параметры генератора
- series_type = "default"  // обычная линейная регрессия

- dim=10 // размерность сигнала
- low=-10, high=10 // границы весов для получения отклика
- noise_var=1 // дисперсия нормального шума
- workers_num=3 // число различных генераторов
- lower_bound=100, upper_bound=400 // границы длин каждого из кусков, получаемых генератором
- alternating=True // нет двух подряд кусков одного типа
#### Фиксированные параметры алгоритма
- from_start=False // начинаем подсчет всех потерь (и составного эксперта) не с самого начала, а после того как прошли по одному разу учатки каждого из  генераторов

- length=1500 // длина изучаемого участка временного ряда

- a=-40, b=40 // параметры алгоритма, отклики отсеиваем тоже по ним

#### Гиперпараметры эксперимента
- windows = [5, 10, 20] // варьируем длину окна обучения ($\frac12x, 1x, 2x$ соответственно, где $x$  &mdash; dim  сигнала )
- weights_function // функция начальных весов, cравниваем $\dfrac1{x^\alpha}$ для разных $\alpha$ и дефолтную $\dfrac{1}{(x+1)\ln^2(x+1)}$
- alpha_function // функция коэффицента $\alpha$ в Mixing Update, cравниваем $\dfrac1{(1 + x)^\alpha}$ для разных $\alpha$

In [34]:
df = pd.read_csv("../results/df-other.csv", index_col=0, sep='\t')
# df.loc[[5, 2, 13, 10, 21, 18]]
df

train_window                weight_function               alpha_function  \
5              5  1 / ((x + 1) * (ln(x + 1))^2)                  1 / (x + 1)   
3              5                  1 / (x^1.001)                  1 / (x + 1)   
7              5                   1 / (x^1.01)                  1 / (x + 1)   
4              5                   1 / (x^1.01)              1 / (x + 1)^1.3   
6              5                   1 / (x^1.01)  1 / ((x + 1)*(ln(x + 1))^2)   
2              5                   1 / (x^1.01)                      1 / e^x   
0              5                   1 / (x^1.01)                  1 / e^(x/3)   
1              5                    1 / (x^1.1)                  1 / (x + 1)   
13            10  1 / ((x + 1) * (ln(x + 1))^2)                  1 / (x + 1)   
11            10                  1 / (x^1.001)                  1 / (x + 1)   
15            10                   1 / (x^1.01)                  1 / (x + 1)   
12            10                   1 / (x^1.01)              1 / (x + 1)^1.3   
14            10                   1 / (x^1.01)  1 / ((x + 1)*(ln(x + 1))^2)   
10            10                   1 / (x^1.01)                      1 / e^x   
8             10                   1 / (x^1.01)                  1 / e^(x/3)   
9             10                    1 / (x^1.1)                  1 / (x + 1)   
21            20  1 / ((x + 1) * (ln(x + 1))^2)                  1 / (x + 1)   
19            20                  1 / (x^1.001)                  1 / (x + 1)   
23            20                   1 / (x^1.01)                  1 / (x + 1)   
20            20                   1 / (x^1.01)              1 / (x + 1)^1.3   
22            20                   1 / (x^1.01)  1 / ((x + 1)*(ln(x + 1))^2)   
18            20                   1 / (x^1.01)                      1 / e^x   
16            20                   1 / (x^1.01)                  1 / e^(x/3)   
17            20                    1 / (x^1.1)                  1 / (x + 1)   

         mean   random_1   random_2   random_3   random_4  
5   235135.48  239444.02  198209.12  267607.19  235281.57  
3   190819.44  190700.95  162997.91  210267.49  199311.42  
7   190184.28  190122.27  162227.69  209655.72  198731.46  
4   166154.00  170966.07  134417.11  185983.37  173249.44  
6   135296.21  165184.09   40253.48  174482.82  161264.46  
2   133224.39  165339.42   37392.26  171568.82  158597.05  
0   133225.94  165340.95   37392.05  171568.85  158601.91  
1   194532.69  195382.82  164837.91  215724.05  202185.99  
13  183257.14  180371.38  162705.30  199049.29  190902.58  
11  129067.96  120891.72  116613.22  134385.77  144381.15  
15  128692.52  120599.04  116009.26  134129.99  144031.81  
12  121927.84  117345.07   99196.28  132689.00  138481.01  
14  132777.02  128680.24  101917.92  147801.88  152708.03  
10  142611.45  136190.72  106754.33  159380.42  168120.32  
8   142614.39  136193.94  106755.27  159382.56  168125.79  
9   134971.88  126964.79  120524.71  142112.21  150285.81  
21  182775.02  177974.00  156669.36  194418.89  202037.85  
19  126549.10  121954.65  108621.81  131318.63  144301.29  
23  126841.61  122328.09  108658.49  131768.57  144611.30  
20  150836.36  149090.84  123139.84  162399.27  168715.51  
22  185658.32  183890.47  150629.07  201688.07  206425.68  
18  214711.72  205743.23  170230.00  230120.91  252752.72  
16  214719.22  205750.56  170234.26  230130.27  252761.78  
17  135185.84  130106.96  115368.10  141036.35  154231.96

In [4]:
df_w = df[df["alpha_function"] == "1 / (x + 1)"].iloc[:, :4]
df_w

train_window                weight_function alpha_function       mean
5              5  1 / ((x + 1) * (ln(x + 1))^2)    1 / (x + 1)  235135.48
3              5                  1 / (x^1.001)    1 / (x + 1)  190819.44
7              5                   1 / (x^1.01)    1 / (x + 1)  190184.28
1              5                    1 / (x^1.1)    1 / (x + 1)  194532.69
13            10  1 / ((x + 1) * (ln(x + 1))^2)    1 / (x + 1)  183257.14
11            10                  1 / (x^1.001)    1 / (x + 1)  129067.96
15            10                   1 / (x^1.01)    1 / (x + 1)  128692.52
9             10                    1 / (x^1.1)    1 / (x + 1)  134971.88
21            20  1 / ((x + 1) * (ln(x + 1))^2)    1 / (x + 1)  182775.02
19            20                  1 / (x^1.001)    1 / (x + 1)  126549.10
23            20                   1 / (x^1.01)    1 / (x + 1)  126841.61
17            20                    1 / (x^1.1)    1 / (x + 1)  135185.84

In [28]:
df_a = df[(df["weight_function"] == "1 / (x^1.01)") & ((df["alpha_function"] == "1 / (x + 1)") | (df["alpha_function"] == "1 / e^x"))]
dfa = df_a[["alpha_function", "mean"]]


print(dfa.to_latex(index=False))

\begin{tabular}{lr}
\toprule
alpha_function & mean \\
\midrule
1 / (x + 1) & 190184.280000 \\
1 / e^x & 133224.390000 \\
1 / (x + 1) & 128692.520000 \\
1 / e^x & 142611.450000 \\
1 / (x + 1) & 126841.610000 \\
1 / e^x & 214711.720000 \\
\bottomrule
\end{tabular}



In [5]:
df[["train_window", "mean"]].groupby("train_window").mean()

mean
train_window               
5             117750.785556
10             72380.342222
20             86215.462222

#### Выводы

##### Train window
В среднем, при окне обучения равном dim=10, результаты оказались лучше чем для размеров 5 и 20.

##### Weights function
При всех рассмотренных окнах, регрет для функции весов $\dfrac1{x^\alpha}$ меньше, чем для дефолтной $\dfrac{1}{(x+1)\ln^2(x+1)}$, и при стремлении $\alpha$ к 1 регрет уменьшается. То есть минимальный регрет среди рассмотренных здесь достигается  для $\dfrac1{x^{1.05}}$, она же была мной зафиксирована для экспериментами над alpha_function.

##### Alpha function
При окнах обучения $5$ и $10$, побеждают $\dfrac1{(x+1)^{1.5}}$ и $\dfrac1{(x+1)^{1.3}}$, но отрыв от остальных не слишком велик. \
При окне $20$ побеждает дефолтная $\dfrac1{(x+1)}$ c нарастающим отрывом при увеличении $\alpha$

**Таким образом, на практике лучшей функцией для начальных весов является $\dfrac1{x^{1+\varepsilon}}$ c как можно меньшим $\varepsilon$**

**Значительных улучшений дефолтной функции $\dfrac1{(x+1)}$ для коэффицента $\alpha$ Mixing Update  пока нет**

Теоретическая оценка сверху выглядит нормально: